File: houserobber_exhaustive.java
//Exhaustive
//TC: Exponential
class Solution {
    public int rob(int[] nums) {
        if(nums == null) return 0;
        return helper(nums, 0, 0);
    }
    
    private int helper(int[] nums, int i, int amount){
        //base
        if(i >= nums.length) return amount;
        
        //logic
        //choose
        int case1 = helper(nums, i+2, amount + nums[i]);
        //not choose
        int case0 = helper(nums, i+1, amount);
        
        int result = Math.max(case1,case0);
        return result;
    }
    
}

File: house-robber-exhaustive.cpp
// Time Complexity : O(2^n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : No since time complexity is exponential which is why for some cases time limit exceeds.

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;


class Solution {
public:
    int rob(vector<int>& nums) {
        return helper(nums, 0, 0);
    }

    int helper(vector<int>& nums, int idx, int robbings){
        //base 
        if(idx >= nums.size()) return robbings;
        //logic
        //choose
        int case1 = helper(nums, idx + 2, robbings + nums[idx]);

        //didnt choose
        int case2 = helper(nums, idx + 1, robbings);

        return max(case1, case2);
    }
};